python/pyspark/sql/streaming.py

```diff
@@ -189,6 +189,282 @@ def resetTerminated(self):
+            return self._jsqs.latency().get()
+        else:
+            return None
+
+    @property
+    @since(2.1)
+    def sourceStatuses(self):
+        """
+        Current statuses of the sources.
+
+        >>> len(sqs.sourceStatuses)
+        1
+        >>> sqs.sourceStatuses[0].description
+        u'MySource1'
+        """
+        return [SourceStatus(ss) for ss in self._jsqs.sourceStatuses()]
+
+    @property
+    @since(2.1)
+    def sinkStatus(self):
+        """
+        Current status of the sink.
+
+        >>> sqs.sinkStatus.description
+        u'MySink'
+        """
+        return SinkStatus(self._jsqs.sinkStatus())
+
+    @property
+    @since(2.1)
+    def triggerStatus(self):
+        """
+        Low-level detailed status of the last completed/currently active trigger.
+
+        >>> sqs.triggerStatus
+        {u'key': u'value'}
+        """
+        return self._jsqs.triggerStatus()
+
+
+class SourceStatus(object):
+    """
+    Status and metrics of a streaming Source.
+
+    .. note:: Experimental
+
+    .. versionadded:: 2.1
+    """
+
+    def __init__(self, jss):
+        self._jss = jss
+
+    def __str__(self):
+        """
+        Pretty string of this source status.
+
+        >>> print(sqs.sourceStatuses[0])
+        SourceStatus:    MySource1
+            Available offset: #0
+            Input rate: 4.0 rows/sec
+            Processing rate: 5.0 rows/sec
+            Trigger status:
+                key: value
+        """
+        return self._jss.toString()
+
+    @property
+    @ignore_unicode_prefix
+    @since(2.1)
+    def description(self):
+        """
+        Description of the source corresponding to this status.
+
+        >>> sqs.sourceStatuses[0].description
+        u'MySource1'
+        """
+        return self._jss.description()
+
+    @property
+    @ignore_unicode_prefix
+    @since(2.1)
+    def offsetDesc(self):
+        """
+        Description of the current offset if known.
+
+        >>> sqs.sourceStatuses[0].offsetDesc
+        u'#0'
+        """
+        return self._jss.offsetDesc()
+
+    @property
+    @since(2.1)
+    def inputRate(self):
+        """
+        Current rate (rows/sec) at which data is being generated by the source.
+
+        >>> sqs.sourceStatuses[0].inputRate
+        4.0
+        """
+        return self._jss.inputRate()
+
+    @property
+    @since(2.1)
+    def processingRate(self):
+        """
+        Current rate (rows/sec) at which the query is processing data from the source.
+
+        >>> sqs.sourceStatuses[0].processingRate
+        5.0
+        """
+        return self._jss.processingRate()
+
+    @property
+    @since(2.1)
+    def triggerStatus(self):
+        """
+        Low-level detailed status of the last completed/currently active trigger.
+
+        >>> sqs.sourceStatuses[0].triggerStatus
+        {u'key': u'value'}
+       """
+        return self._jss.triggerStatus()
+
+
+class SinkStatus(object):
```

Comment:
No OutputRate? (Though you did mention in the comments that it doesn't work with whole-stage codegen.